text/3753-rest-pattern-matches-non-struct.md

```diff
@@ -0,0 +1,138 @@
+```rust
+// The library's API is later changed to this.
+pub enum Foo {
+    A,
+    B,
+}
+```
+
+Intuitively, one might think that this change is backwards-compatible. However,
+this is technically not the case, since client code might match the `Foo` type
+with a rest pattern (`Foo { .. }`), which currently only matches structs, and
+not enums or other types.
+```rust
+// Client code that uses the library
+// Works with the initial API, but doesn't work with the later API.
+fn do_something(x: Foo) {
+    match x {
+        Foo { .. } => {}
+    }
+}
+```
+
+To eliminate this semver hazard, this RFC proposes that the pattern `Foo { .. }`
+should match values of any type named `Foo`, not just structs.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+As a special case, a struct pattern which contain only a rest pattern, and no
+other fields, can match with any value of the appropriate type, even if it is
+not a struct.
+
+For example, the pattern `Foo { .. }` can match with any value that have type
+`Foo`, even if `Foo` is not a struct (e.g., it might be an enum).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+(The following text is appended to the section on [struct
+patterns](https://doc.rust-lang.org/stable/reference/patterns.html#struct-patterns)
+in the reference.)
+
+As a special case, if a struct pattern contains only `..` as the fields, and the
+path refers to a type (as opposed to an enum variant), then the pattern is an
+irrefutable pattern that matches against any value of that type. This applies
+even if that type is not a struct type (e.g., it might be an enum type, or it
+might be a type alias, etc.)
+
+For example, the pattern `foo::Bar::<Baz> { .. }` can match any value of the
+type `foo::Bar<Baz>` (even if this type is not a struct type), or match a value
+of the type `foo<Baz>` that contains the enum variant `Bar`.
+
+Formally, this special case applies to the following syntax: *PathInExpression*
+`{` *StructPatternEtCetera* `}`
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This adds a complication to Rust. 
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+* This special case is only applied to `Foo { .. }`, and not `Foo(..)`.
+  * This is because, currently, the pattern `Foo(..)` can only match a tuple
+  struct whose fields are all public, so the `Foo(..)` pattern does not pose a
+  semver hazard.
+  * On the other hand, the pattern `Foo { .. }` currently matches any struct
+  named `Foo`, including tuple structs, and including type aliases that refer to
+  structs. This makes the meaning of the pattern `Foo { .. }` already similar to
+  "match any type named `Foo`".
+* As an alternative, we could deprecate the pattern `Foo { .. }` (either in all
+  cases, or only in cases where `Foo` has no public fields). We could then
+  potentially remove this pattern from the language in a future edition.
```

Comment:
What are the use cases for using the `Foo { .. }` pattern when `Foo` is a struct with only private fields?  If we were to start linting against that, what would we be losing?